stdlib/public/core/StringBreadcrumbs.swift

```diff
@@ -0,0 +1,139 @@
+    let stride = _StringBreadcrumbs.breadcrumbStride
+
+    self.crumbs = []
+
+    if str.isEmpty {
+      self.utf16Length = 0
+      return
+    }
+
+    self.crumbs.reserveCapacity(
+      (str._guts.count / 3) / stride)
+
+    // TODO(String performance): More efficient implementation of initial scan.
+    // We'll also want to benchmark this initial scan in order to track changes.
+
+    let utf16 = str.utf16
+    var i = 0
+    var curIdx = utf16.startIndex
+    while curIdx != utf16.endIndex {
+      if i % stride == 0 { //i.isMultiple(of: stride) {
+        self.crumbs.append(curIdx)
+      }
+      i = i &+ 1
+      curIdx = utf16.index(after: curIdx)
+    }
+
+    // Corner case: index(_:offsetBy:) can produce the endIndex
+    if i % stride == 0 {
+      self.crumbs.append(utf16.endIndex)
+    }
+
+    self.utf16Length = i
+    _sanityCheck(self.crumbs.count == 1 + (self.utf16Length / stride))
+
+    _invariantCheck()
+  }
+}
+
+extension _StringBreadcrumbs {
+  var stride: Int {
+    @inline(__always) get { return _StringBreadcrumbs.breadcrumbStride }
+  }
+
+  // Fetch the lower-bound index corresponding to the given offset, returning
+  // the index and the remaining offset to adjust
+  internal func getBreadcrumb(
+    forOffset offset: Int
+  ) -> (lowerBound: String.Index, remaining: Int) {
+    return (crumbs[offset / stride], offset % stride)
+  }
+
+  // Fetch the lower-bound offset corresponding to the given index, returning
+  // the lower-bound and its offset
+  internal func getBreadcrumb(
+    forIndex idx: String.Index
+  ) -> (lowerBound: String.Index, offset: Int) {
+    var lowerBound = idx.encodedOffset / 3 / stride
+    var upperBound = Swift.min(1 + (idx.encodedOffset / stride), crumbs.count)
+    _sanityCheck(crumbs[lowerBound] <= idx)
+    _sanityCheck(upperBound == crumbs.count || crumbs[upperBound] >= idx)
+
+    while (upperBound &- lowerBound) > 1 {
+      let mid = lowerBound + ((upperBound &- lowerBound) / 2)
+      if crumbs[mid] <= idx { lowerBound = mid } else { upperBound = mid }
+    }
+
+    let crumb = crumbs[lowerBound]
+    _sanityCheck(crumb <= idx)
+    _sanityCheck(lowerBound == crumbs.count-1 || crumbs[lowerBound+1] > idx)
+
+    return (crumb, lowerBound &* stride)
+  }
+
+  #if !INTERNAL_CHECKS_ENABLED
+  @nonobjc @inline(__always) internal func _invariantCheck() {}
+  #else
+  @nonobjc @inline(never) @_effects(releasenone)
+  internal func _invariantCheck() {
+  }
+  #endif // INTERNAL_CHECKS_ENABLED
+}
+
+extension _StringGuts {
+  @_effects(releasenone)
+  internal func getBreadcrumbsPtr() -> UnsafePointer<_StringBreadcrumbs> {
+    _sanityCheck(hasBreadcrumbs)
+
+    let mutPtr: UnsafeMutablePointer<_StringBreadcrumbs?>
+    if hasNativeStorage {
+      mutPtr = _object.nativeStorage._breadcrumbsAddress
+    } else {
+      mutPtr = UnsafeMutablePointer(
+        Builtin.addressof(&_object.sharedStorage._breadcrumbs))
+    }
+
+    if _slowPath(mutPtr.pointee == nil) {
```

Comment:
Shouldn't this use an atomic load?

Reply:
I don't think it has to be. `populateBreadcrumbs` constructs a new instance and atomic compare-and-exchanges it. Worst case there will be redundant work that will be thrown away at the end. That's rare and probably a better price to pay than an atomic operation on every single read of the breadcrumbs.

CC @atrick, thoughts?

Reply:
I don't think it has to be. populateBreadcrumbs constructs a new instance and atomic compare-and-exchanges it. Worst case there will be redundant work that will be thrown away at the end. That's rare and probably a better price to pay than an atomic operation on every single read of the breadcrumbs.

CC @atrick, thoughts?

Reply:
Given that the breadcrumbs are immutable data that are published once and
remain valid, and that breadcrumb data is published via a pointer
value that is atomically written after initializing the data....

Then we definitely don't want to do any hardware atomic operations on
the read of the breadcrumbs. In theory, this load does need to be a
"memory_order_consume", but I think a comment would suffice. (It would
be ridiculous if LLVM stopped assuming that all loads had "consume"
semantics).

To prove safety, we just need these conditions:

- A happens-before relationship exists between the store of the
  breadcrumb values and the store of the publishing pointer. Since
  _stdlib_atomicInitializeARCRef is the only way to write the
  _StringBreadcrumbs reference inside the storage, that's
  covered. Minimally, an atomic store-release is
  required. cmpxchg_seqcst_seqcst is actually overkill.

- The only way to read a breadcrumb is via a dependent load on the
  published _StringBreadcrumbs reference (consume ordering).

- Nothing stores to the breadcrumb data again after it is
  published. You could potentially allow atomic relaxed consistency
  stores to the breadcumb later if the order the changes are seen didn't
  matter. Then the reads would also need to be atomic relaxed
  consistency, which in practice should not cost anything but would
  prevent word-tearing. In your case that wouldn't make sense anyway.

It would be a good idea to comment these assumptions, in some more concise form.

Reply:
Perhaps we should make an AtomicInitializableRef construct around an UMP to capture this usecase, which occurs a few times elsewhere in the stdlib. Array/Set/Dictionary do something very similar to implement deferred bridging, except they use atomic loads — they too would benefit from a smarter implementation.